c10/util/Exception.h

```diff
@@ -112,17 +113,56 @@ class C10_API Error : public std::exception {
   std::string compute_what(bool include_backtrace) const;
 };
 
+class C10_API Warning {
+ public:
+  Warning(
+      const SourceLocation& source_location,
+      const std::string& msg,
+      bool verbatim);
+
+  Warning(SourceLocation source_location, const char* msg, bool verbatim);
+
+  Warning(
+      SourceLocation source_location,
+      ::c10::detail::CompileTimeEmptyString msg,
+      bool verbatim);
+
+  virtual ~Warning();
+
+  // Getters for members
+  const SourceLocation& source_location() const;
+  const std::string& msg() const;
+  bool verbatim() const;
+
+ private:
+  // Where the warning happened.
+  SourceLocation source_location_;
+
+  // The actual warning message.
+  std::string msg_;
+
+  // See note: [Verbatim Warnings]
+  bool verbatim_;
+};
+
+class C10_API UserWarning : public Warning {
+  using Warning::Warning;
+};
+
+typedef c10::variant<UserWarning> warning_variant_t;
```

Comment:
I'm not a specialist here, what is the benefit of using variant here instead of just the base class of which everything should be inheriting from?

Reply:
If we just used the base class, then when we copy a warning into a `vector<Warning>`, we lose the information about which subclass it was

Reply:
But the virtual methods on it will still work fine right? We don't need the type to actually be correct right?

Reply:
Unfortunately, no. Here's an example:

<details>
<summary>Click to expand</summary>

```cpp
#include <iostream>
#include <vector>

class Base {
public:
  virtual int id() {
    return 0;
  }
};

class Derived : public Base {
public:
  int id() override {
    return 1;
  }
};

int main() {
  std::vector<Base> buffer;

  Base base;
  Derived derived;

  buffer.push_back(base);
  buffer.push_back(derived);

  std::cout << buffer[0].id() << std::endl;
  std::cout << buffer[1].id() << std::endl;
}
```

</summary>

Running the above code, both `buffer[...].id()` calls return 0, meaning that the base method was called both times

Reply:
Ho right, can we make these unique_ptr then? :D

Reply:
Yes, smart pointers are an option. There was some discussion earlier in this PR about whether to use smart pointers or variants.

Originally, I did implement this PR with `shared_ptr` (though `unique_ptr` would have been a better choice), and it did work. But one of the the annoying things about it was that to create a `shared_ptr` of the right type you have to call `std::make_shared<subclass_type>` (or `std::make_unique<subclass_type` for `unique_ptr`), where the "subclass_type" here has to actually be the actual subclass type, not the base type. Because of that, in order to have a generic way to create a unique_ptr from any of the warning subclasses, you have to make a template function with explicit overloads for each warning subclass, something like this (code might not be 100% correct):

```cpp
template<>
std::shared_ptr warning_to_shared_ptr(c10::UserWarning warning) {
  return std::make_shared<c10::UserWarning>(warning);
}

// repeat for each warning subclass type
```

Perhaps there was a better way to do it, I'm not sure.

If smart pointers are actually better, then I'm not opposed to switching back. But at the moment it seems to me like using variants makes the implementation simpler.

cc @lezcano

Reply:
My point here is, why do you need to know the type of the objects / why do you need virtual things at all? If they all have a given method, can't you just do a whatever's the equivalent we have to `std::visit` and execute that method for all the elements in the `std::vector`?

Reply:
> why do you need to know the type of the objects / why do you need virtual things at all?

I think we have three main options:

1. Keep track of the `c10::Warning` subclass type so that we can map it to the appropriate Python warning type later (whether that's done by the explicit typeid thing currently in this PR or the better way Alban mentioned here: https://github.com/pytorch/pytorch/pull/84101#discussion_r982516484).

2. Add an enum member to `c10::Warning` that we can use to map to the appropriate Python warning type later.

3. Just keep track of the Python warning type directly as a `PyObject*` member of `c10::Warning`, with no need for subclasses.

I went for option 1 because it seemed more consistent with how `c10::Error` works, and it's consistent with the Python side, where warning types are all subclasses of a base class.

Option 2 is mentioned here https://github.com/pytorch/pytorch/pull/84101#issuecomment-1245878625, and I'm not sure if it's better or not, because it seems less consistent with `c10::Error` and Python warnings.

Option 3 is might be the simplest, but I'm not sure how to implement it, and again it seems less consistent.

> If they all have a given method, can't you just do a whatever's the equivalent we have to std::visit and execute that method for all the elements in the std::vector?

I'm having trouble understanding this. What is an example of something that's equivalent to std::visit?

Reply:
Consistency with `c10::Error` and python side of things is a good reason for sure.
I am not familiar with the variant approach that's why I was asking. But sounds good.

Reply:
I think either of the 3 options should work. I liked the variant option because it avoids pointers, which is good. Now, I think that @amjames is looking into this, so he'll be helping him finalise this PR :)